test/org/javarosa/xpath/expr/ToDateTest.java

```diff
@@ -2,113 +2,136 @@
-    // NaN is xpath's 'null values'
-    assertEquals(Double.NaN, toDate(Double.NaN, false));
-    assertEquals(Double.NaN, toDate(Double.NaN, true));
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void doubleValuesLessThanTheIntegerMinThrow() {
-    toDate(Integer.valueOf(Integer.MIN_VALUE).doubleValue() - 1, false);
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void doubleValuesGreaterThanTheIntegerMaxThrow() {
-    toDate(Integer.valueOf(Integer.MAX_VALUE).doubleValue() + 1, false);
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void positiveInfinityThrows() {
-    toDate(POSITIVE_INFINITY, false);
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void negativeInfinityThrows() {
-    toDate(NEGATIVE_INFINITY, false);
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void unparseableDateStringsThrow() {
-    toDate("some random text", false);
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void booleansThrow() {
-    // We test this type specifically because, according to the documentation
-    // Dates can be encoded as booleans, but booleans can't be decoded into Dates
-    toDate(false, false);
-  }
-
-  @Test(expected = XPathTypeMismatchException.class)
-  public void anyOtherTypeThrows() {
-    // We will test just one type to cover the final 'else' block.
-    // We can't explicitly test for all possible types.
-    toDate(2L, false);
-  }
+
+    private static Date date(int year, int month, int day) {
+        return date(LocalDate.of(year, month, day).atStartOfDay().atZone(ZoneId.systemDefault()));
+    }
+
+    private static Date date(int year, int month, int day, int hour, int minute, int second, int milli) {
+        return date(LocalDateTime.of(year, month, day, hour, minute, second, milli * 1_000_000).atZone(ZoneId.systemDefault()));
+    }
+
+    private static Date date(ZonedDateTime zonedDateTime) {
+        return Date.from(zonedDateTime.toInstant());
+    }
+
+    private static Date date(OffsetDateTime offsetDateTime) {
+        return Date.from(offsetDateTime.toInstant());
+    }
+
+    @Test
+    public void convertsISO8601DatesWithoutPreservingTime() {
+        assertEquals(
+                date(2018, 1, 1),
+                toDate("2018-01-01", false)
+        );
+    }
+
+    @Test
+    public void convertsISO8601DatesWithoutOffsetPreservingTime() {
+        assertEquals(
+                date(2018, 1, 1, 10, 20, 30, 400),
+                toDate("2018-01-01T10:20:30.400", true)
+        );
+    }
+
+    @Test
+    public void convertsISO8601DatesWithOffsetPreservingTime() {
+        assertEquals(
+                date(OffsetDateTime.of(2018, 1, 1, 10, 20, 30, 400_000_000, ZoneOffset.ofHours(2))),
+                toDate("2018-01-01T10:20:30.400+02", true)
+        );
+    }
+
+    @Test
+    public void convertsTimestampsWithoutPreservingTime() {
+        assertEquals(
```

Comment:
I get this:

java.lang.AssertionError: 
Expected :Thu Dec 31 00:00:00 PST 1970
Actual   :Fri Jan 01 00:00:00 PST 1971
 <Click to see difference>
	at org.javarosa.xpath.expr.ToDateTest.convertsTimestampsWithoutPreservingTime(ToDateTest.java:64)

Reply:
Good catch, thanks! Time zones are hard!